### ISSUE TITLE
Follow-Up to Issue 112 - Custom Element renders TR/TD outside of table

### ISSUE THREAD

**Issue by jlgreene2** [aurelia/binding#279]
I am attempting to implement a table which displays a tree structure of data using recursive custom elements.  Unfortunately, because the HTML parser is so specific about what tags can be included in the TABLE / TBODY / TR tags, the output of the custom element ends up being rendered outside the table because my custom element isn't considered valid by the parser.  

In one of Rob's comments to this issue, the idea of implementing a new syntax (<tr element="template"></tr>) was discussed.  The comment stated that this feature wouldn't be incorporated before the beta.  I couldn't find any follow-up discussion on this feature, and was wondering if it was still being considered?  I'm kind of dead in the water with respect to attempting to implement this capability into my application, without having to resort to low-level DOM coding.  

This seems like it would be a pretty important feature to have, to enable the binding / templating  framework to be capable of implement anything beyond fairly simple HTML tables.  Thoughts?

**Comment by EisenbergEffect** [comment]
It's under consideration. But, the easiest thing to do would be to not use a table. Tables have rendering issues and even the modern data-grids, for example, often don't use tables at all.

**Comment by jlgreene2** [comment]
Rob, 

Thank you very much for your reply.

From the research that I have done, Infragistics IgniteUI and Telerik Kendo controls (who appear to be the two market leaders in the commercial controls space at this point in time) both render HTML tables for their grids.  They are not using DIV/CSS-based grids.  I am by no means a web development expert, so my perspective may be a bit uninformed.  But building grids using CSS's and DIV's seems to make things like colspan and rowspan very difficult to get right, and these are features that are very common in LOB applications rendering significant quantities of back-office data.

I really like where Aurelia is going, and what I'm attempting to do is determine whether there's a path to using Aurelia completely free of these third party grid controls.  I believe Aurelia can make it possible to create very capable, light-weight data grids with very little code, which ultimately my developers will have the ability to completely control how they are rendered.  But, I think the issue of not being able to embed custom elements in tables to perform the type of recursion I'm attempting could be limiting.

Again, I can't say enough about the quality of the Aurelia framework.  I've been evaluating it for the last month of so, purposefully looking for edge cases where it might not fit our needs.  This is really the only issue I've come across after more than thirty days of putting it through its paces.

**Comment by EisenbergEffect** [comment]
You can embed custom elements, you just can't embed non-spec-compliant html, so where you put the elements matter. You can't use one in place of a tr or td, because that's not valid html. The browser parsers literally see those elements and mangle them before we can do anything about it.

In general, I would recommend that you find another way. Even if we were to come up with some workaround..it would be a hack. Aside from not being spec-compliant today, it also wouldn't align with the Web Components specification, which has been an important design goal for Aurelia.

I need a bit more concrete information about what you are trying to do to guide you further. I've got some ideas, but I'm not sure what would work for you without more details.

**Comment by jlgreene2** [comment]
Hypothetically, lets say I was wanting to write a set of custom elements representing a generic data grid, which supports features like grouping by one or more attributes.  The data that I want to display looks something like this.

```
var data = [
	{
	group:  "Group 1", 
    data : [
    	{
      	group:	"Group 1.1", 
        data:  [
        	{
          	"column1":  "foo", 
            "column2":  "bar", 
            "column3":  "baz"
          }
        ]
      }, 
      {
      	group:	"Group 1.2", 
        data: [
        	{
          	"column1":	"foo", 
            "column2": 	"bar", 
            "column3": 	"baz"
          }
        ]
      }, 
      {
      	group:	"Group 1.3", 
        data: {
          group:	"Group 1.3.1", 
          data: [
          	{
            	"column1":	"foo", 
            	"column2": 	"bar", 
            	"column3": 	"baz"
            }, 
            {
            	"column1":	"foo", 
            	"column2": 	"bar", 
            	"column3": 	"baz"            
            }, 
            {
            	"column1":	"foo", 
            	"column2": 	"bar", 
            	"column3": 	"baz"            
            }
          ]
        }
      }
    ]
  }, 
];
```

Because I want the grid control to be generic, and capable of rendering just about any type of source data, I also need to provide some configuration data regarding all of the columns that will be displayed in the table.  The configuration might look something like this:

```
var configuration = {
	columns: [
  	{
      fieldName:	"column1", 
      columnLabel:  "Column 1"
    }, 
    {
      fieldName: "column2", 
      columnLabel:  "Column 2"
    }, 
    {
      fieldName:  "column3", 
      columnLabel:  "Column 3"
    }  
  ]
}
```

Essentially, I want to create a table that looks like one of the two examples shown here:  https://jsfiddle.net/jlgreene2/s3q9Lyxg/ 

I went down the path of creating a custom element for the table object itself.  Here's a simplified version of the view model and HTML file.

```
// View Model
import {customElement, bindable} from 'aurelia-framework';
import _ from "lodash";

@customElement('simplegrid')
export class simplegrid { 
    @bindable columns
    @bindable data
    ...
}

// HTML file
<template>
    <table border="1" width="100%">
        <thead>
            <tr>
                <th repeat.for="column of columns">${column.columnLabel}</th>
            </tr>
        </thead>
        <tbody repeat.for="current of data">           
                <!-- Invoke a second custom element to render the columns of data -->
                <datarow columns.bind="columns" row.bind="current"></datarow>
            </tr>
        </tbody>
    </table>
</template>
```

Here was the concept of the datarow custom element:

```
// View Model
import {customElement, containerless, bindable} from 'aurelia-framework';
import _ from "lodash";

@customElement('datarow')
@containerless
export class DataRpw { 
    @bindable columns
    @bindable row
    ...
}

// HTML File
<template>
    <tr>
        <td colspan.bind="columns.length">${row.group}</td>
    </tr>
    <tr repeat.for="line of row.data">
        <td repeat.for="column of columns">${line[column.fieldName]}</td>
    </tr>
    <datarow repeat.for="child of row.data" columns.bind="columns" data.bind="child"></datarow>
</template>
```

This can't work, because the datarow custom element isn't a valid HTML tag within the tbody tag, and gets puked out early by the HTML parser.  What I need to be able to do is to render one or more table rows as a combined unit, and to recursively display all levels of the data hierarchy for the source data.

Does this help explain what I'm trying to accomplish?  I really appreciate you taking the time to look at this.

**Comment by jlgreene2** [comment]
I found a solution to my problem.  I took the hierarchical data structure I referenced above and flattened it into array of objects (using Lodash and some recursion) that contain types that are attributed in such as way that I can differentiate between group-level rows (requiring COLSPAN's) and data level rows containing the actual columns to be displayed.  I set up a repeater on the TBODY tag of my table to iterate over this new array, and used conditional bindings to create TR rows representing either group-level rows or data level rows.  It seems to work for what I was attempting to accomplish.
